### PR TITLE
Add OpenMP to hashmap

### PR DESCRIPTION
**Pull request by ksuarez1423**
# Summary

- This PR is a feature
- It does the following:
  - Modifies previously created hash map to support OpenMP execution.
  - Ports tests for this version. 
  - Will be followed by PR adding GPU accelerator support.

Continued development for #574

### CODE REVIEW

File: src/axom/core/Map.hpp [hunk 1]
```diff
@@ -0,0 +1,834 @@
...
+ *  RAJA and Umpire to allow use of map within kernels. Resembles 
+ *  unordered_map, which we can't use due to a lack of host-device decoration.
+ *
+ * \tparam Key the type of keys. Must allow equality comparison, must be copyable. 
+ * \tparam T the type of values to hold, must be copyable.
+ * \tparam Hash functor that takes an object of Key type and returns 
+ *  a hashed value of size_t (for now)
+ */
+
+template <typename Key, typename T, typename Hash = std::hash<Key>>
```
Comment (kennyweiss) [inline src/axom/core/Map.hpp:353]: Looking good @ksuarez1423 

Quick early feedback -- I think we'll want to add a `Policy` (or `ExecPolicy`) parameter so the use can explicitly request the execution policy. E.g. in an `OpenMP` configuration of axom, we might still want a sequential hashmap.

Comment (ksuarez1423) [reply src/axom/core/Map.hpp]: Implemented, as well as exactly what I needed -- I had been trying to work out the best way to handle guarding things that needed to be guarded, and this answered that perfectly.

File: src/axom/core/Map.hpp [hunk 2]
```diff
@@ -491,23 +500,31 @@ class Map
    */
   axom_map::Pair<Key, T> insert(const Key& key, const T& val)
   {
-    //If branching becomes an isssue, this protective statement will be 
-    //removed in lieu of expecting the user to not try to insert after 
+    //If branching becomes an isssue, this protective statement will be
```
Comment (agcapps) [inline src/axom/core/Map.hpp:503]: typo "isssue"

Comment (rhornung67) [reply src/axom/core/Map.hpp:503]: "eagle eye" capps scores again!

File: src/axom/core/Map.hpp [hunk 3]
```diff
@@ -746,7 +860,10 @@ class Map
   float m_load_factor; /*!< currently unused value, used in STL unordered_map to determine when to resize, which we don't do internally at the moment */
   axom_map::Node<Key, T> m_end; /*!< the node with meaningless values allowing for user to verify success or failure of operations */
```
Comment (agcapps) [inline src/axom/core/Map.hpp:861]: I'd rephrase the documenting comment to something like "the sentinel node allowing the user to verify success of operations"

File: src/axom/core/Map.hpp [hunk 4]
```diff
@@ -746,7 +860,10 @@ class Map
   float m_load_factor; /*!< currently unused value, used in STL unordered_map to determine when to resize, which we don't do internally at the moment */
```
Comment (agcapps) [inline src/axom/core/Map.hpp:860]: Typo in Doxygen for `m_size` --- "currenty"

File: src/axom/core/Map.hpp [hunk 5]
```diff
@@ -742,11 +856,14 @@ class Map
   axom_map::Bucket<Key, T>* m_buckets; /*!< array of pointers to linked lists containing data */
   std::size_t m_bucket_count; /*!< the number of buckets in the Map instance */
   int m_bucket_len; /*!< the number of items that can be contained in a bucket in this Map instance */
-  int m_size; /*!< the number of items currenty stored in this Map instance */
+  int m_size; /*!< the number of items currently stored in this Map instance */
   float m_load_factor; /*!< currently unused value, used in STL unordered_map to determine when to resize, which we don't do internally at the moment */
-  axom_map::Node<Key, T> m_end; /*!< the node with meaningless values allowing for user to verify success or failure of operations */
+  axom_map::Node<Key, T> m_end; /*!< the sentinel node enabling verification of operation success or failure */
```
Comment (kennyweiss) [inline src/axom/core/Map.hpp:861]: 👍

File: src/axom/core/tests/CMakeLists.txt [hunk 6]
```diff
@@ -89,6 +89,29 @@ if (ENABLE_MPI)
...
+                      HEADERS    ${core_mpi_tests}
+                      OUTPUT_DIR ${TEST_OUTPUT_DIRECTORY}
+                      DEPENDS_ON ${utils_tests_depends} openmp
+                      FOLDER     axom/core/tests )
+
+  foreach( test_suite ${core_openmp_tests} )
+    get_filename_component( test_suite ${test_suite} NAME_WE )
+    axom_add_test( NAME          ${test_suite}_openmp
+                   COMMAND       core_openmp_tests --gtest_filter=${test_suite}*
+                   NUM_MPI_TASKS 1 )
```
Comment (kennyweiss) [inline src/axom/core/tests/CMakeLists.txt:111]: I think we have a way to specify the number of omp threads too

File: src/axom/core/tests/core_openmp_main.cpp [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+#include "core_openmp_map.hpp"
+
+int main(int argc, char** argv)
+{
+  ::testing::InitGoogleTest(&argc, argv);
+
+  // add this line to avoid a warning in the output about thread safety
+  ::testing::FLAGS_gtest_death_test_style = "threadsafe";
+
+#ifdef EIGEN_SOLVE_TESTER_SHOULD_SEED
```
Comment (kennyweiss) [inline src/axom/core/tests/core_openmp_main.cpp:19]: This seems like a copy/paste artifact.
``EIGEN_SOLVE_TESTER_SHOULD_SEED`` will never be true in this file.

If we should test this using seeds, we need a `#define` in this file that is set at the top of this file
e.g. `OMP_MAP_TESTER_SHOULD_SEED` ?

File: src/axom/core/tests/core_openmp_map.hpp [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    0,
+    test.max_size(),
+    AXOM_LAMBDA(IndexType idx) {
+      Key key = idx;
+      T value = key * 27;
+      auto ret_test = test2->insert(key, value);
+      EXPECT_EQ(true, ret_test.second);
+    });
+  EXPECT_EQ(false, test.empty());
+  axom::for_all<axom::OMP_EXEC>(
```
Comment (kennyweiss) [inline src/axom/core/tests/core_openmp_map.hpp:42]: Please refactor the tests in this file w.r.t. the policy: 
They should *either* use the templated `Policy` *or* be explicitly written for `axom::OMP_EXEC`.
As is, it seems to be in a middle ground -- having a `Policy` but not using it.

File: src/axom/core/tests/core_openmp_map.hpp [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+      Key key = idx;
+      EXPECT_EQ(key * 27, test2->find(key).value);
+    });
+  //This should fail, since we're at capacity.
+  auto ret = test.insert(test.max_size(), 900);
+  EXPECT_EQ(false, ret.second);
+}
+
+template <typename Key, typename T, typename Hash, typename Policy>
+void test_brackets(experimental::Map<Key, T, Hash, Policy> &test)
```
Comment (kennyweiss) [inline src/axom/core/tests/core_openmp_map.hpp:55]: Suggestion: `brackets` ->`subscript` ?

File: src/axom/core/tests/core_openmp_map.hpp [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+{
+  experimental::Map<Key, T, Hash, Policy> *test2 = &test;
+  axom::for_all<axom::OMP_EXEC>(
+    0,
+    test.size(),
+    AXOM_LAMBDA(IndexType idx) {
+      Key key = idx;
+      EXPECT_EQ(key * 27, (*test2)[key]);
+    });
```
Comment (kennyweiss) [inline src/axom/core/tests/core_openmp_map.hpp:64]: I don't see any reason why the following wouldn't work, 
but consider also expressing this as:

```cpp
experimental::Map<Key, T, Hash, Policy>& test2 = test;
```
which would allow for expressions like:
```cpp
EXPECT_EQ(key * 27, test[key]);
```

Comment (kennyweiss) [reply src/axom/core/tests/core_openmp_map.hpp:64]: It might also be good to express this as the following to help make it more obvious what you're testing 
```cpp
  Key key = idx;
  Value actual_val = test2[key];  // or Value& ?
  Value exp_val = key*27;
  EXPECT_EQ(exp_val, actual_val);
```

**Review by kennyweiss — APPROVED** [review]
This is really nice @ksuarez1423 

If possible, please restore `const` to accessor functions like `operator[]` and `find()`

Preparing for the next step, please think about how to restructure the tests so that the same tests can be used for multiple policies (`SEQ`, `OMP`, `CUDA`, ...)

I also have two questions I should have asked on the previous PR:
* What is the difference between `Map::insert()` and `Map::insert_or_assign()`? Please make sure this distinction comes across clearly in the docs.  (I'm probably missing something obvious).
* Is `bucket()` ever called without `get_key()`? If not, would an overload of `bucket` that takes a `Key` be helpful?

Please also make sure to:
- [ ] format the code using `make style`
- [ ] ensure there are no warnings
- [ ] update the release-notes

File: src/axom/core/tests/CMakeLists.txt (hunk 6 above)
Comment (kennyweiss) [reply src/axom/core/tests/CMakeLists.txt:111]: `NUM_OMP_THREADS`

https://github.com/LLNL/axom/blob/57a6ef7ab50e113e4cf4b639657eb84ff10789c0/src/cmake/AxomMacros.cmake#L118-L127

See also: https://llnl-blt.readthedocs.io/en/develop/api/target.html#blt-add-test

**Review by agcapps — APPROVED** [review]
Kae, this is a solid development toward an accelerated hashmap.  Thanks for your conscientious work!

File: src/axom/core/tests/core_openmp_main.cpp (hunk 7 above)
Comment (agcapps) [reply src/axom/core/tests/core_openmp_main.cpp:19]: +1.  Thanks; I missed this.

**Review by rhornung67 — APPROVED** [review]

File: src/axom/core/Map.hpp [hunk 11]
```diff
@@ -491,23 +500,31 @@ class Map
    */
   axom_map::Pair<Key, T> insert(const Key& key, const T& val)
   {
-    //If branching becomes an isssue, this protective statement will be 
-    //removed in lieu of expecting the user to not try to insert after 
+    //If branching becomes an issue, this protective statement will be
+    //removed in lieu of expecting the user to not try to insert after
     //clear().
```
[the remark this answers is not in the document]
Comment (ksuarez1423) [reply src/axom/core/Map.hpp:505]: Micro-optimization thoughts about how the GPU dislikes branching, that's all.

File: src/axom/core/Map.hpp [hunk 12]
```diff
@@ -563,7 +589,7 @@ class Map
    * \return A const reference to the value of key-value pair in the Map instance that is associated with supplied Key.
    *
    */
-  const T& operator[](const Key& key) const
+  const T& operator[](const Key& key)
```
[the remark this answers is not in the document]
Comment (ksuarez1423) [reply src/axom/core/Map.hpp:592]: Arlie actually caught this, and I forgot. I'll get that in tomorrow.

File: src/axom/core/Map.hpp [hunk 13]
```diff
@@ -734,6 +769,85 @@ class Map
     return hashed;
   }
 
+  //NOTE: Every single locking function will use the OpenMP locks if Axom was compiled with OpenMP support,
+  //regardless of whether the user code is actually using it.
```
[the remark this answers is not in the document]
Comment (ksuarez1423) [reply src/axom/core/Map.hpp:773]: The comment is not at all applicable, I'll be removing it.

**Comment by ksuarez1423** [comment]
> This is really nice @ksuarez1423
> 
> If possible, please restore `const` to accessor functions like `operator[]` and `find()`
> 
> Preparing for the next step, please think about how to restructure the tests so that the same tests can be used for multiple policies (`SEQ`, `OMP`, `CUDA`, ...)
> 
> I also have two questions I should have asked on the previous PR:
> 
> * What is the difference between `Map::insert()` and `Map::insert_or_assign()`? Please make sure this distinction comes across clearly in the docs.  (I'm probably missing something obvious).
> * Is `bucket()` ever called without `get_key()`? If not, would an overload of `bucket` that takes a `Key` be helpful?
> 
> Please also make sure to:
> 
> * [ ]  format the code using `make style`
> * [ ]  ensure there are no warnings
> * [ ]  update the release-notes

In the brief description of insert_or_assign(), it has the additional wording "or updates value of existing key-value pair if item with supplied key already exists." insert() lacks this wording. Should insert have the negative added? Something like "will not update value of existing key-value pairs"?

As for bucket(), I just checked, and that it needs the hash actually goes against the standard. I'll make it call with a key, and internally call get_hash(). 

The warnings are a difficult step. My overloading to ensure the correct locking functions are called will always send a warning, because the policy argument goes un-used. Using said argument in a meaningful way isn't happening, because there is no meaningful way to use it.

**Comment by kennyweiss** [comment]
@ksuarez1423 -- in case you haven't seen it, `AXOM_UNUSED_VAR` is intended to elide warnings for variables that are used in some code paths but not others, e.g. within an `#ifdef`

See:
https://github.com/LLNL/axom/blob/57a6ef7ab50e113e4cf4b639657eb84ff10789c0/src/axom/core/Macros.hpp#L143-L163

File: src/axom/core/Map.hpp [hunk 14]
```diff
@@ -778,15 +778,23 @@ class Map
    * \param [in] index the index needed for locking, if this weren't sequential.
    * \param [in] overload execution space object for the sake of function overloading.
    */
-  void bucket_lock(std::size_t index, axom::SEQ_EXEC overload) const { (void)index; (void) overload; } 
+  void bucket_lock(std::size_t index, axom::SEQ_EXEC overload) const
+  {
+    (void)index;
+    (void)overload;
```
Comment (kennyweiss) [inline src/axom/core/Map.hpp:784]: Please change these to:
```suggestion
    AXOM_UNUSED_VAR(index);
    AXOM_UNUSED_VAR(overload);
```
